libs/core/synchronization/include/hpx/synchronization/mutex.hpp

```diff
@@ -32,75 +33,398 @@ namespace hpx { namespace threads {
+    ///        \a mutex offers exclusive, non-recursive ownership semantics:
+    ///
+    ///        - A calling thread owns a mutex from the time that it
+    ///          successfully calls either \a lock or \a try_lock until it
+    ///          calls \a unlock.
+    ///        - When a thread owns a \a mutex, all other threads will
+    ///          block (for calls to \a lock) or receive a \a false return
+    ///          value (for \a try_lock) if they attempt to claim ownership
+    ///          of the \a mutex.
+    ///        - A calling thread must not own the \a mutex prior to
+    ///          calling \a lock or \a try_lock.
+    ///
+    ///        The behavior of a program is undefined if a \a mutex is
+    ///        destroyed while still owned by any threads, or a thread
+    ///        terminates while owning a \a mutex. The mutex class satisfies
+    ///        all requirements of \a Mutex and \a StandardLayoutType.
+    ///
+    ///        \a hpx::mutex is neither copyable nor movable.
+    ///
     class mutex
     {
     public:
+        /// \brief \a hpx::mutex is neither copyable nor movable
         HPX_NON_COPYABLE(mutex);
 
     protected:
+        /// \cond NOPROTECTED
         typedef hpx::spinlock mutex_type;
+        /// \endcond NOPROTECTED
 
     public:
+        ///
+        /// \brief Constructs the \a mutex. The \a mutex is in unlocked state
+        ///        after the constructor completes.
+        ///
+        ///       FIXME(bhumit->hkaiser): true for hpx::mutex?
+        /// \note Because the default constructor is \a constexpr, static
+        ///       mutexes are initialized as part of static non-local
+        ///       initialization, before any dynamic non-local initialization
+        ///       begins. This makes it safe to lock a \a mutex in a constructor
+        ///       of any static object.
+        ///
+        /// \param description description of the \a mutex.
+        ///
         HPX_CORE_EXPORT mutex(char const* const description = "");
 
+        ///
+        /// \brief Destroys the \a mutex.
+        ///        The behavior is undefined if the \a mutex is owned by any
+        ///        thread or if any thread terminates while holding any
+        ///        ownership of the \a mutex.
+        ///
         HPX_CORE_EXPORT ~mutex();
 
-        HPX_CORE_EXPORT void lock(
-            char const* description, error_code& ec = throws);
+        ///
+        /// \brief Locks the mutex. If another thread has already locked the
+        ///        mutex, a call to lock will block execution until the lock is
+        ///        acquired.
+        ///        If lock is called by a thread that already owns the mutex,
+        ///        the behavior is undefined: for example, the program may
+        ///        deadlock.
+        ///        FIXME(bhumit->hkaiser) do we follow either of the following?
+        ///        An implementation that can detect the invalid
```

Comment:
@hkaiser does our implementation follow either of these?

Reply:
Yes, this is true for HPX as well.